### DISCUSSION TITLE
Ordering is a bit goofy

### DISCUSSION THREAD

**Discussion by jimkohl** [LefterisXris/CodeTour#81] [Q&A] [answered]
Hi - your tool was suggested at a DX session @ Github Universe this year, so I downloaded it and am putting it through its paces.

If I want to create say 3 tours - beginner, advanced, expert - after creating them the order I created them in won't be kept in the tool window.  doesn't appear to be alphabetical or even order I added it; maybe its the value of the id hash that is ordering.  

Either way that's not ideal - and I cannot find a way to reorder these in the plugin UX or by editing the source files for each tour.
Any suggestions?
Jim

**Comment by LefterisXris** [comment, marked as answer]
Hello @jimkohl, that's a good point! 

Currently, tours are loaded either from IntelliJ indexed **VFS - Virtual File System** so the order depends on it. I am not really sure what is the order that is used there. It seems random, or based on a hash as you said.

Order seems to be important as you mentioned, so yes, I am considering providing a couple of options for this. A quick thought is to have by default `alphabetical` order and provide options (from settings) for `date-added` and `manual` ordering both with `ASC/DESC` options.
I am going to further design this feature, and proceed to a release shortly. It will be a really useful feature for everyone, especially for cases when someone has to deal with lots of Tours.

Thanks for pointing it out! I'll keep you updated!  ✅

**Reply by jimkohl** [reply]
Great yes in the code I could see that the file system determines loading.  I think for now you could simplify this a bit and simply sort alphabetically - and then each tour could be prefixed with an integer like "1 Introduction".  Not so sure ASC/DESC is all that useful here as normally we curate the order and we are set.  

Alternatively you could assign a metadata property called 'order' per tour, which would allow titles to be arranged non-alphabetically and you could drag/drop the order.

**Reply by LefterisXris** [reply]
Yes, my thoughts are almost the same. Check the issue I created for more details: https://github.com/LefterisXris/CodeTour/issues/82

**Reply by jimkohl** [reply]
These options are good - but in lieue of having a full blown implementation of those, could we at least have some ordering mechanism that is not relying on the file system?  Just something whether it is alphabeticall or metadata.  

really like this tool but misordered tours cause confusion for consumers and I want to advocate for others to use this tool!

**Reply by LefterisXris** [reply]
Sure @jimkohl, what you are saying is fair and probably _required_ for this tool, so stay tuned. I will do it on the weekend.

**Reply by LefterisXris** [reply]
Hello @jimkohl. Ordering features (https://github.com/LefterisXris/CodeTour/issues/82) has been included in latest release: https://plugins.jetbrains.com/plugin/19227-codetour

Release info: https://github.com/LefterisXris/CodeTour/releases/tag/v0.0.10
More details in the related PR: https://github.com/LefterisXris/CodeTour/pull/90

Once again, thanks for the feedback. I am going to continue to the next features

**Reply by jimkohl** [reply]
looks great already verified that alpha ordering works.  tx

**Reply by muescha** [reply]
With the following idea the alphabetical / and or other sort methods are not needed: 

1) Insert point for new steps

- new steps are inserted at the end of a tour as a new step. 
- if a user has selected a step at the left, then new steps comes after the selected step. the new step is selected and following steps inserted also after this step.

2) changing order: just drag and drop the steps up and down to get a new order. maybe it should be possible to select more than one step for the reorder.

**Reply by LefterisXris** [reply]
Hi @muescha! 
Good points, thanks for the feedback! However, your suggestions are for Steps, while the initial ordering discussion was for the Tours. 

Regarding the order of Steps of a Tour, I agree it could be more enhanced. For now you can change the order using `right-click` and then  `move up` or `move down` options, as shown in the screenshot.

![image](https://github.com/LefterisXris/CodeTour/assets/22836249/beb9ab93-2ed7-486d-8994-fea2ba4ff796)

It may not be optimal from UX perspective, but it's fully functional. 
I will consider **multiple selection** and **drag&drop** for changing Steps' order in future releases!

**Reply by muescha** [reply]
I'd suggest using a similar ordering system for the tours. Alphabetical ordering might work if I label them with prefixes like 1. and 2., but that feels more like a workaround. Ordering by date might not be ideal either, especially when inserting a new tour among older ones becomes complicated.

**Reply by LefterisXris** [reply]
Ordering of Tours has already been implemented. By default it is `alphabetical` but user can change it to sort by date. 
Check the screenshot here for details: https://github.com/LefterisXris/CodeTour/pull/90

**Reply by muescha** [reply]
I'm saying that there's no necessity to define an order and uphold a setup screen when users can simply adjust the order through drag-and-drop or by using the context menu. This approach requires less code to maintain (no plugin setup screen).

**Comment by jimkohl** [comment]
Alpha sorting and a manual mode (using drag/drop) should both be implemented; and the same approach should be provided for any lists (tour and event lists).  I think the alpha mode was the easiest to setup which is why it was suggested that to begin with.  

Both approaches are useful - alpha ordering works using labelling and it isn't a workaround.  
It has a benefit in that the content at rest on the file system is ordered.

there is a nice app I use for managing music setlists for my bands called forscore.  can sort by name or manually.
For an existing collection of music in a new set - I'll use manual ordering to get the flow of tunes just right.
But if I'm attempting to find something - using alpha sorting is easiest.  

If I'm building a bespoke training (like in this situation) I'd be more inclined to use naming for ordering, although drag/drop would be nice.   w naming if order changes, then need to change the prefixes.  Not sure which requires less code tho.
